toolchain/check/call.cpp

```diff
@@ -9,60 +9,100 @@
 #include "toolchain/check/convert.h"
 #include "toolchain/check/deduce.h"
 #include "toolchain/check/function.h"
+#include "toolchain/sem_ir/entity_with_params_base.h"
 #include "toolchain/sem_ir/ids.h"
 #include "toolchain/sem_ir/inst.h"
 #include "toolchain/sem_ir/typed_insts.h"
 
 namespace Carbon::Check {
 
+// Information about a callee in a checked call to an entity.
+struct CalleeInfo {
+  // Whether the call appears to be valid.
+  bool is_valid;
+  // The specific version of the callee to invoke.
+  SemIR::SpecificId specific_id;
+};
+
+static auto ResolveCalleeInCall(Context& context, SemIR::LocId loc_id,
+                                const SemIR::EntityWithParamsBase& entity,
+                                llvm::StringLiteral entity_kind_for_diagnostic,
+                                SemIR::GenericId entity_generic_id,
+                                SemIR::SpecificId enclosing_specific_id,
+                                SemIR::InstId self_id,
+                                llvm::ArrayRef<SemIR::InstId> arg_ids)
+    -> CalleeInfo {
+  CalleeParamsInfo callee_info(entity);
+
+  // Check that the arity matches.
+  auto params = context.inst_blocks().GetOrEmpty(callee_info.param_refs_id);
+  if (arg_ids.size() != params.size()) {
+    CARBON_DIAGNOSTIC(CallArgCountMismatch, Error,
+                      "{0} argument(s) passed to {1} expecting "
+                      "{2} argument(s).",
+                      int, llvm::StringLiteral, int);
+    CARBON_DIAGNOSTIC(InCallToEntity, Note, "Calling {0} declared here.",
+                      llvm::StringLiteral);
+    context.emitter()
+        .Build(loc_id, CallArgCountMismatch, arg_ids.size(),
+               entity_kind_for_diagnostic, params.size())
+        .Note(callee_info.callee_loc, InCallToEntity,
+              entity_kind_for_diagnostic)
+        .Emit();
+    return {.is_valid = false, .specific_id = SemIR::SpecificId::Invalid};
+  }
+
+  // Perform argument deduction.
+  auto specific_id = SemIR::SpecificId::Invalid;
+  if (entity_generic_id.is_valid()) {
+    specific_id = DeduceGenericCallArguments(
+        context, loc_id, entity_generic_id, enclosing_specific_id,
+        callee_info.implicit_param_refs_id, callee_info.param_refs_id, self_id,
+        arg_ids);
+    if (!specific_id.is_valid()) {
+      return {.is_valid = false, .specific_id = SemIR::SpecificId::Invalid};
```

Comment:
Had you considered a `std::optional<SpecificId>` return, with `nullopt` for invalid calls?

Reply:
Done.